### PR TITLE
V2: Sourcemap Fixes

### PR DESCRIPTION
**Pull request by DeMoorJasper**
# ↪️ Pull Request

This PR fixes various issues with Parcel 2 sourcemaps:
- Fix sourcemap comment having incorrect format
- Add endpoint to dev server for the fancy hosted sources (this works a lot better than I expected)
- Fix issues with offsets

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

File: packages/core/integration-tests/test/sourcemaps.js [hunk 1]
```diff
@@ -64,54 +64,210 @@ function checkSourceMapping({
...
+    /*checkSourceMapping({
```
Comment (DeMoorJasper) [inline packages/core/integration-tests/test/sourcemaps.js]: Not sure why this fails, the map seems pretty correct when I visualise it.
Perhaps an issue with the checkSourceMapping util?

cc @mischnic you have any clue why this happens?

**Comment by mischnic** [comment]
Another issue with `inlineSources`: [this](https://sokra.github.io/source-map-visualization/#base64,Ly8gbW9kdWxlcyBhcmUgZGVmaW5lZCBhcyBhbiBhcnJheQovLyBbIG1vZHVsZSBmdW5jdGlvbiwgbWFwIG9mIHJlcXVpcmVzIF0KLy8KLy8gbWFwIG9mIHJlcXVpcmVzIGlzIHNob3J0IHJlcXVpcmUgbmFtZSAtPiBudW1lcmljIHJlcXVpcmUKLy8KLy8gYW55dGhpbmcgZGVmaW5lZCBpbiBhIHByZXZpb3VzIGJ1bmRsZSBpcyBhY2Nlc3NlZCB2aWEgdGhlCi8vIG9yaWcgbWV0aG9kIHdoaWNoIGlzIHRoZSByZXF1aXJlIGZvciBwcmV2aW91cyBidW5kbGVzCgovLyBlc2xpbnQtZGlzYWJsZS1uZXh0LWxpbmUgbm8tZ2xvYmFsLWFzc2lnbgpwYXJjZWxSZXF1aXJlID0gZnVuY3Rpb24obW9kdWxlcywgY2FjaGUsIGVudHJ5LCBnbG9iYWxOYW1lKSB7CiAgLy8gU2F2ZSB0aGUgcmVxdWlyZSBmcm9tIHByZXZpb3VzIGJ1bmRsZSB0byB0aGlzIGNsb3N1cmUgaWYgYW55CiAgdmFyIHByZXZpb3VzUmVxdWlyZSA9IHR5cGVvZiBwYXJjZWxSZXF1aXJlID09PSAnZnVuY3Rpb24nICYmIHBhcmNlbFJlcXVpcmU7CiAgdmFyIG5vZGVSZXF1aXJlID0gdHlwZW9mIHJlcXVpcmUgPT09ICdmdW5jdGlvbicgJiYgcmVxdWlyZTsKCiAgZnVuY3Rpb24gbmV3UmVxdWlyZShuYW1lLCBqdW1wZWQpIHsKICAgIGlmICghY2FjaGVbbmFtZV0pIHsKICAgICAgaWYgKCFtb2R1bGVzW25hbWVdKSB7CiAgICAgICAgLy8gaWYgd2UgY2Fubm90IGZpbmQgdGhlIG1vZHVsZSB3aXRoaW4gb3VyIGludGVybmFsIG1hcCBvcgogICAgICAgIC8vIGNhY2hlIGp1bXAgdG8gdGhlIGN1cnJlbnQgZ2xvYmFsIHJlcXVpcmUgaWUuIHRoZSBsYXN0IGJ1bmRsZQogICAgICAgIC8vIHRoYXQgd2FzIGFkZGVkIHRvIHRoZSBwYWdlLgogICAgICAgIHZhciBjdXJyZW50UmVxdWlyZSA9CiAgICAgICAgICB0eXBlb2YgcGFyY2VsUmVxdWlyZSA9PT0gJ2Z1bmN0aW9uJyAmJiBwYXJjZWxSZXF1aXJlOwogICAgICAgIGlmICghanVtcGVkICYmIGN1cnJlbnRSZXF1aXJlKSB7CiAgICAgICAgICByZXR1cm4gY3VycmVudFJlcXVpcmUobmFtZSwgdHJ1ZSk7CiAgICAgICAgfQoKICAgICAgICAvLyBJZiB0aGVyZSBhcmUgb3RoZXIgYnVuZGxlcyBvbiB0aGlzIHBhZ2UgdGhlIHJlcXVpcmUgZnJvbSB0aGUKICAgICAgICAvLyBwcmV2aW91cyBvbmUgaXMgc2F2ZWQgdG8gJ3ByZXZpb3VzUmVxdWlyZScuIFJlcGVhdCB0aGlzIGFzCiAgICAgICAgLy8gbWFueSB0aW1lcyBhcyB0aGVyZSBhcmUgYnVuZGxlcyB1bnRpbCB0aGUgbW9kdWxlIGlzIGZvdW5kIG9yCiAgICAgICAgLy8gd2UgZXhoYXVzdCB0aGUgcmVxdWlyZSBjaGFpbi4KICAgICAgICBpZiAocHJldmlvdXNSZXF1aXJlKSB7CiAgICAgICAgICByZXR1cm4gcHJldmlvdXNSZXF1aXJlKG5hbWUsIHRydWUpOwogICAgICAgIH0KCiAgICAgICAgLy8gVHJ5IHRoZSBub2RlIHJlcXVpcmUgZnVuY3Rpb24gaWYgaXQgZXhpc3RzLgogICAgICAgIGlmIChub2RlUmVxdWlyZSAmJiB0eXBlb2YgbmFtZSA9PT0gJ3N0cmluZycpIHsKICAgICAgICAgIHJldHVybiBub2RlUmVxdWlyZShuYW1lKTsKICAgICAgICB9CgogICAgICAgIHZhciBlcnIgPSBuZXcgRXJyb3IoIkNhbm5vdCBmaW5kIG1vZHVsZSAnIiArIG5hbWUgKyAiJyIpOwogICAgICAgIGVyci5jb2RlID0gJ01PRFVMRV9OT1RfRk9VTkQnOwogICAgICAgIHRocm93IGVycjsKICAgICAgfQoKICAgICAgbG9jYWxSZXF1aXJlLnJlc29sdmUgPSByZXNvbHZlOwogICAgICBsb2NhbFJlcXVpcmUuY2FjaGUgPSB7fTsKCiAgICAgIHZhciBtb2R1bGUgPSAoY2FjaGVbbmFtZV0gPSBuZXcgbmV3UmVxdWlyZS5Nb2R1bGUobmFtZSkpOwoKICAgICAgbW9kdWxlc1tuYW1lXVswXS5jYWxsKAogICAgICAgIG1vZHVsZS5leHBvcnRzLAogICAgICAgIGxvY2FsUmVxdWlyZSwKICAgICAgICBtb2R1bGUsCiAgICAgICAgbW9kdWxlLmV4cG9ydHMsCiAgICAgICAgdGhpcwogICAgICApOwogICAgfQoKICAgIHJldHVybiBjYWNoZVtuYW1lXS5leHBvcnRzOwoKICAgIGZ1bmN0aW9uIGxvY2FsUmVxdWlyZSh4KSB7CiAgICAgIHJldHVybiBuZXdSZXF1aXJlKGxvY2FsUmVxdWlyZS5yZXNvbHZlKHgpKTsKICAgIH0KCiAgICBmdW5jdGlvbiByZXNvbHZlKHgpIHsKICAgICAgcmV0dXJuIG1vZHVsZXNbbmFtZV1bMV1beF0gfHwgeDsKICAgIH0KICB9CgogIGZ1bmN0aW9uIE1vZHVsZShtb2R1bGVOYW1lKSB7CiAgICB0aGlzLmlkID0gbW9kdWxlTmFtZTsKICAgIHRoaXMuYnVuZGxlID0gbmV3UmVxdWlyZTsKICAgIHRoaXMuZXhwb3J0cyA9IHt9OwogIH0KCiAgbmV3UmVxdWlyZS5pc1BhcmNlbFJlcXVpcmUgPSB0cnVlOwogIG5ld1JlcXVpcmUuTW9kdWxlID0gTW9kdWxlOwogIG5ld1JlcXVpcmUubW9kdWxlcyA9IG1vZHVsZXM7CiAgbmV3UmVxdWlyZS5jYWNoZSA9IGNhY2hlOwogIG5ld1JlcXVpcmUucGFyZW50ID0gcHJldmlvdXNSZXF1aXJlOwogIG5ld1JlcXVpcmUucmVnaXN0ZXIgPSBmdW5jdGlvbihpZCwgZXhwb3J0cykgewogICAgbW9kdWxlc1tpZF0gPSBbCiAgICAgIGZ1bmN0aW9uKHJlcXVpcmUsIG1vZHVsZSkgewogICAgICAgIG1vZHVsZS5leHBvcnRzID0gZXhwb3J0czsKICAgICAgfSwKICAgICAge30KICAgIF07CiAgfTsKCiAgZm9yICh2YXIgaSA9IDA7IGkgPCBlbnRyeS5sZW5ndGg7IGkrKykgewogICAgbmV3UmVxdWlyZShlbnRyeVtpXSk7CiAgfQoKICBpZiAoZW50cnkubGVuZ3RoKSB7CiAgICAvLyBFeHBvc2UgZW50cnkgcG9pbnQgdG8gTm9kZSwgQU1EIG9yIGJyb3dzZXIgZ2xvYmFscwogICAgLy8gQmFzZWQgb24gaHR0cHM6Ly9naXRodWIuY29tL0ZvcmJlc0xpbmRlc2F5L3VtZC9ibG9iL21hc3Rlci90ZW1wbGF0ZS5qcwogICAgdmFyIG1haW5FeHBvcnRzID0gbmV3UmVxdWlyZShlbnRyeVtlbnRyeS5sZW5ndGggLSAxXSk7CgogICAgLy8gQ29tbW9uSlMKICAgIGlmICh0eXBlb2YgZXhwb3J0cyA9PT0gJ29iamVjdCcgJiYgdHlwZW9mIG1vZHVsZSAhPT0gJ3VuZGVmaW5lZCcpIHsKICAgICAgbW9kdWxlLmV4cG9ydHMgPSBtYWluRXhwb3J0czsKCiAgICAgIC8vIFJlcXVpcmVKUwogICAgfSBlbHNlIGlmICh0eXBlb2YgZGVmaW5lID09PSAnZnVuY3Rpb24nICYmIGRlZmluZS5hbWQpIHsKICAgICAgZGVmaW5lKGZ1bmN0aW9uKCkgewogICAgICAgIHJldHVybiBtYWluRXhwb3J0czsKICAgICAgfSk7CgogICAgICAvLyA8c2NyaXB0PgogICAgfSBlbHNlIGlmIChnbG9iYWxOYW1lKSB7CiAgICAgIHRoaXNbZ2xvYmFsTmFtZV0gPSBtYWluRXhwb3J0czsKICAgIH0KICB9CgogIC8vIE92ZXJyaWRlIHRoZSBjdXJyZW50IHJlcXVpcmUgd2l0aCB0aGlzIG5ldyBvbmUKICByZXR1cm4gbmV3UmVxdWlyZTsKfSh7IjBhZDEzYzVhYjM3OTRiMzM0MmJlNmZkOWMxZDhmYjE1IjpbZnVuY3Rpb24ocmVxdWlyZSxtb2R1bGUsZXhwb3J0cykgewpjb25zdCBsb2NhbCA9IHJlcXVpcmUoJy4vbG9jYWwnKTsKCm1vZHVsZS5leHBvcnRzID0gZnVuY3Rpb24gKCkgewogIHJldHVybiBsb2NhbC5hICsgbG9jYWwuYjsKfTsKfSx7Ii4vbG9jYWwiOiI2MDhiZDhmMWUwMmJkYTRkZTBlZjA3NzAwODgxMzIyMiJ9XSwiNjA4YmQ4ZjFlMDJiZGE0ZGUwZWYwNzcwMDg4MTMyMjIiOltmdW5jdGlvbihyZXF1aXJlLG1vZHVsZSxleHBvcnRzKSB7CmNvbnN0IHV0aWwgPSByZXF1aXJlKCcuL3V0aWxzL3V0aWwnKTsKCmV4cG9ydHMuYSA9IDU7CmV4cG9ydHMuYiA9IHV0aWwuY291bnQoNCwgNSk7Cn0seyIuL3V0aWxzL3V0aWwiOiJiNWMzMWUwY2NlYTI1NDJjNTI4NWRkM2U3OTU4YTE0ZiJ9XSwiYjVjMzFlMGNjZWEyNTQyYzUyODVkZDNlNzk1OGExNGYiOltmdW5jdGlvbihyZXF1aXJlLG1vZHVsZSxleHBvcnRzKSB7CmV4cG9ydHMuY291bnQgPSBmdW5jdGlvbiAoYSwgYikgewogIHJldHVybiBhICsgYjsKfTsKfSx7fV19LHt9LFsiMGFkMTNjNWFiMzc5NGIzMzQyYmU2ZmQ5YzFkOGZiMTUiXSwgbnVsbCkKCi8vIyBzb3VyY2VNYXBwaW5nVVJMPWluZGV4LmpzLm1hcAo=,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInBhY2thZ2VzL2NvcmUvaW50ZWdyYXRpb24tdGVzdHMvdGVzdC9pbnRlZ3JhdGlvbi9zb3VyY2VtYXAtbmVzdGVkL2luZGV4LmpzIiwicGFja2FnZXMvY29yZS9pbnRlZ3JhdGlvbi10ZXN0cy90ZXN0L2ludGVncmF0aW9uL3NvdXJjZW1hcC1uZXN0ZWQvbG9jYWwuanMiLCJwYWNrYWdlcy9jb3JlL2ludGVncmF0aW9uLXRlc3RzL3Rlc3QvaW50ZWdyYXRpb24vc291cmNlbWFwLW5lc3RlZC91dGlscy91dGlsLmpzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiI7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7OztBQUFBO0FBQ0E7QUFDQTtBQUNBO0FBQ0E7QUFDQTtBQUNBLEFDTkE7QUFDQTtBQUNBO0FBQ0E7QUFDQTtBQUNBLEFDTEE7QUFDQTtBQUNBO0FBQ0E7QUFDQSIsImZpbGUiOiJpbmRleC5qcy5tYXAiLCJzb3VyY2VSb290IjoiLi4vLi4vLi4vLi4vLi4vLi4vLi4vIiwic291cmNlc0NvbnRlbnQiOlsiXCIwYWQxM2M1YWIzNzk0YjMzNDJiZTZmZDljMWQ4ZmIxNVwiOltmdW5jdGlvbihyZXF1aXJlLG1vZHVsZSxleHBvcnRzKSB7XG5jb25zdCBsb2NhbCA9IHJlcXVpcmUoJy4vbG9jYWwnKTtcblxubW9kdWxlLmV4cG9ydHMgPSBmdW5jdGlvbiAoKSB7XG4gIHJldHVybiBsb2NhbC5hICsgbG9jYWwuYjtcbn07XG59LHtcIi4vbG9jYWxcIjpcIjYwOGJkOGYxZTAyYmRhNGRlMGVmMDc3MDA4ODEzMjIyXCJ9XSIsIixcIjYwOGJkOGYxZTAyYmRhNGRlMGVmMDc3MDA4ODEzMjIyXCI6W2Z1bmN0aW9uKHJlcXVpcmUsbW9kdWxlLGV4cG9ydHMpIHtcbmNvbnN0IHV0aWwgPSByZXF1aXJlKCcuL3V0aWxzL3V0aWwnKTtcblxuZXhwb3J0cy5hID0gNTtcbmV4cG9ydHMuYiA9IHV0aWwuY291bnQoNCwgNSk7XG59LHtcIi4vdXRpbHMvdXRpbFwiOlwiYjVjMzFlMGNjZWEyNTQyYzUyODVkZDNlNzk1OGExNGZcIn1dIiwiLFwiYjVjMzFlMGNjZWEyNTQyYzUyODVkZDNlNzk1OGExNGZcIjpbZnVuY3Rpb24ocmVxdWlyZSxtb2R1bGUsZXhwb3J0cykge1xuZXhwb3J0cy5jb3VudCA9IGZ1bmN0aW9uIChhLCBiKSB7XG4gIHJldHVybiBhICsgYjtcbn07XG59LHt9XSJdfQ==,IjBhZDEzYzVhYjM3OTRiMzM0MmJlNmZkOWMxZDhmYjE1IjpbZnVuY3Rpb24ocmVxdWlyZSxtb2R1bGUsZXhwb3J0cykgewpjb25zdCBsb2NhbCA9IHJlcXVpcmUoJy4vbG9jYWwnKTsKCm1vZHVsZS5leHBvcnRzID0gZnVuY3Rpb24gKCkgewogIHJldHVybiBsb2NhbC5hICsgbG9jYWwuYjsKfTsKfSx7Ii4vbG9jYWwiOiI2MDhiZDhmMWUwMmJkYTRkZTBlZjA3NzAwODgxMzIyMiJ9XQ==,LCI2MDhiZDhmMWUwMmJkYTRkZTBlZjA3NzAwODgxMzIyMiI6W2Z1bmN0aW9uKHJlcXVpcmUsbW9kdWxlLGV4cG9ydHMpIHsKY29uc3QgdXRpbCA9IHJlcXVpcmUoJy4vdXRpbHMvdXRpbCcpOwoKZXhwb3J0cy5hID0gNTsKZXhwb3J0cy5iID0gdXRpbC5jb3VudCg0LCA1KTsKfSx7Ii4vdXRpbHMvdXRpbCI6ImI1YzMxZTBjY2VhMjU0MmM1Mjg1ZGQzZTc5NThhMTRmIn1d,LCJiNWMzMWUwY2NlYTI1NDJjNTI4NWRkM2U3OTU4YTE0ZiI6W2Z1bmN0aW9uKHJlcXVpcmUsbW9kdWxlLGV4cG9ydHMpIHsKZXhwb3J0cy5jb3VudCA9IGZ1bmN0aW9uIChhLCBiKSB7CiAgcmV0dXJuIGEgKyBiOwp9Owp9LHt9XQ==) 
This is part of the emitted source:
```js
"0ad13c5ab3794b3342be6fd9c1d8fb15":[function(require,module,exports) {
```
(not sure if this is related to the other issue)

**Comment by DeMoorJasper** [comment]
@mischnic I haven't written any tests for inlineSources yet, but that's def something we need to fix.

Not entirely sure why it does that though, as I thought it would do the same as non inlined sources. This branch fixes some offset issues though so if you created that with v2 branch it's probably fixed in this one.

File: packages/core/integration-tests/test/sourcemaps.js (hunk 1 above)
Comment (mischnic) [reply packages/core/integration-tests/test/sourcemaps.js]: The "undefined" shouldn't be there
Bildschirmfoto 2019-06-09 um 14 12 38


These duplicate mappings are the issue:
Bildschirmfoto 2019-06-09 um 14 13 55

Comment (DeMoorJasper) [reply packages/core/integration-tests/test/sourcemaps.js]: Yeah you're first comment actually pushed me in the right direction :) Thanks for the quick reply.

I think it's correct now, was basing lineOffset on wrapped. Which was a bad idea as input !== wrapped.

anyways here's the new map: [here](https://sokra.github.io/source-map-visualization/#base64,Ly8gbW9kdWxlcyBhcmUgZGVmaW5lZCBhcyBhbiBhcnJheQovLyBbIG1vZHVsZSBmdW5jdGlvbiwgbWFwIG9mIHJlcXVpcmVzIF0KLy8KLy8gbWFwIG9mIHJlcXVpcmVzIGlzIHNob3J0IHJlcXVpcmUgbmFtZSAtPiBudW1lcmljIHJlcXVpcmUKLy8KLy8gYW55dGhpbmcgZGVmaW5lZCBpbiBhIHByZXZpb3VzIGJ1bmRsZSBpcyBhY2Nlc3NlZCB2aWEgdGhlCi8vIG9yaWcgbWV0aG9kIHdoaWNoIGlzIHRoZSByZXF1aXJlIGZvciBwcmV2aW91cyBidW5kbGVzCgovLyBlc2xpbnQtZGlzYWJsZS1uZXh0LWxpbmUgbm8tZ2xvYmFsLWFzc2lnbgpwYXJjZWxSZXF1aXJlID0gZnVuY3Rpb24obW9kdWxlcywgY2FjaGUsIGVudHJ5LCBnbG9iYWxOYW1lKSB7CiAgLy8gU2F2ZSB0aGUgcmVxdWlyZSBmcm9tIHByZXZpb3VzIGJ1bmRsZSB0byB0aGlzIGNsb3N1cmUgaWYgYW55CiAgdmFyIHByZXZpb3VzUmVxdWlyZSA9IHR5cGVvZiBwYXJjZWxSZXF1aXJlID09PSAnZnVuY3Rpb24nICYmIHBhcmNlbFJlcXVpcmU7CiAgdmFyIG5vZGVSZXF1aXJlID0gdHlwZW9mIHJlcXVpcmUgPT09ICdmdW5jdGlvbicgJiYgcmVxdWlyZTsKCiAgZnVuY3Rpb24gbmV3UmVxdWlyZShuYW1lLCBqdW1wZWQpIHsKICAgIGlmICghY2FjaGVbbmFtZV0pIHsKICAgICAgaWYgKCFtb2R1bGVzW25hbWVdKSB7CiAgICAgICAgLy8gaWYgd2UgY2Fubm90IGZpbmQgdGhlIG1vZHVsZSB3aXRoaW4gb3VyIGludGVybmFsIG1hcCBvcgogICAgICAgIC8vIGNhY2hlIGp1bXAgdG8gdGhlIGN1cnJlbnQgZ2xvYmFsIHJlcXVpcmUgaWUuIHRoZSBsYXN0IGJ1bmRsZQogICAgICAgIC8vIHRoYXQgd2FzIGFkZGVkIHRvIHRoZSBwYWdlLgogICAgICAgIHZhciBjdXJyZW50UmVxdWlyZSA9CiAgICAgICAgICB0eXBlb2YgcGFyY2VsUmVxdWlyZSA9PT0gJ2Z1bmN0aW9uJyAmJiBwYXJjZWxSZXF1aXJlOwogICAgICAgIGlmICghanVtcGVkICYmIGN1cnJlbnRSZXF1aXJlKSB7CiAgICAgICAgICByZXR1cm4gY3VycmVudFJlcXVpcmUobmFtZSwgdHJ1ZSk7CiAgICAgICAgfQoKICAgICAgICAvLyBJZiB0aGVyZSBhcmUgb3RoZXIgYnVuZGxlcyBvbiB0aGlzIHBhZ2UgdGhlIHJlcXVpcmUgZnJvbSB0aGUKICAgICAgICAvLyBwcmV2aW91cyBvbmUgaXMgc2F2ZWQgdG8gJ3ByZXZpb3VzUmVxdWlyZScuIFJlcGVhdCB0aGlzIGFzCiAgICAgICAgLy8gbWFueSB0aW1lcyBhcyB0aGVyZSBhcmUgYnVuZGxlcyB1bnRpbCB0aGUgbW9kdWxlIGlzIGZvdW5kIG9yCiAgICAgICAgLy8gd2UgZXhoYXVzdCB0aGUgcmVxdWlyZSBjaGFpbi4KICAgICAgICBpZiAocHJldmlvdXNSZXF1aXJlKSB7CiAgICAgICAgICByZXR1cm4gcHJldmlvdXNSZXF1aXJlKG5hbWUsIHRydWUpOwogICAgICAgIH0KCiAgICAgICAgLy8gVHJ5IHRoZSBub2RlIHJlcXVpcmUgZnVuY3Rpb24gaWYgaXQgZXhpc3RzLgogICAgICAgIGlmIChub2RlUmVxdWlyZSAmJiB0eXBlb2YgbmFtZSA9PT0gJ3N0cmluZycpIHsKICAgICAgICAgIHJldHVybiBub2RlUmVxdWlyZShuYW1lKTsKICAgICAgICB9CgogICAgICAgIHZhciBlcnIgPSBuZXcgRXJyb3IoIkNhbm5vdCBmaW5kIG1vZHVsZSAnIiArIG5hbWUgKyAiJyIpOwogICAgICAgIGVyci5jb2RlID0gJ01PRFVMRV9OT1RfRk9VTkQnOwogICAgICAgIHRocm93IGVycjsKICAgICAgfQoKICAgICAgbG9jYWxSZXF1aXJlLnJlc29sdmUgPSByZXNvbHZlOwogICAgICBsb2NhbFJlcXVpcmUuY2FjaGUgPSB7fTsKCiAgICAgIHZhciBtb2R1bGUgPSAoY2FjaGVbbmFtZV0gPSBuZXcgbmV3UmVxdWlyZS5Nb2R1bGUobmFtZSkpOwoKICAgICAgbW9kdWxlc1tuYW1lXVswXS5jYWxsKAogICAgICAgIG1vZHVsZS5leHBvcnRzLAogICAgICAgIGxvY2FsUmVxdWlyZSwKICAgICAgICBtb2R1bGUsCiAgICAgICAgbW9kdWxlLmV4cG9ydHMsCiAgICAgICAgdGhpcwogICAgICApOwogICAgfQoKICAgIHJldHVybiBjYWNoZVtuYW1lXS5leHBvcnRzOwoKICAgIGZ1bmN0aW9uIGxvY2FsUmVxdWlyZSh4KSB7CiAgICAgIHJldHVybiBuZXdSZXF1aXJlKGxvY2FsUmVxdWlyZS5yZXNvbHZlKHgpKTsKICAgIH0KCiAgICBmdW5jdGlvbiByZXNvbHZlKHgpIHsKICAgICAgcmV0dXJuIG1vZHVsZXNbbmFtZV1bMV1beF0gfHwgeDsKICAgIH0KICB9CgogIGZ1bmN0aW9uIE1vZHVsZShtb2R1bGVOYW1lKSB7CiAgICB0aGlzLmlkID0gbW9kdWxlTmFtZTsKICAgIHRoaXMuYnVuZGxlID0gbmV3UmVxdWlyZTsKICAgIHRoaXMuZXhwb3J0cyA9IHt9OwogIH0KCiAgbmV3UmVxdWlyZS5pc1BhcmNlbFJlcXVpcmUgPSB0cnVlOwogIG5ld1JlcXVpcmUuTW9kdWxlID0gTW9kdWxlOwogIG5ld1JlcXVpcmUubW9kdWxlcyA9IG1vZHVsZXM7CiAgbmV3UmVxdWlyZS5jYWNoZSA9IGNhY2hlOwogIG5ld1JlcXVpcmUucGFyZW50ID0gcHJldmlvdXNSZXF1aXJlOwogIG5ld1JlcXVpcmUucmVnaXN0ZXIgPSBmdW5jdGlvbihpZCwgZXhwb3J0cykgewogICAgbW9kdWxlc1tpZF0gPSBbCiAgICAgIGZ1bmN0aW9uKHJlcXVpcmUsIG1vZHVsZSkgewogICAgICAgIG1vZHVsZS5leHBvcnRzID0gZXhwb3J0czsKICAgICAgfSwKICAgICAge30KICAgIF07CiAgfTsKCiAgZm9yICh2YXIgaSA9IDA7IGkgPCBlbnRyeS5sZW5ndGg7IGkrKykgewogICAgbmV3UmVxdWlyZShlbnRyeVtpXSk7CiAgfQoKICBpZiAoZW50cnkubGVuZ3RoKSB7CiAgICAvLyBFeHBvc2UgZW50cnkgcG9pbnQgdG8gTm9kZSwgQU1EIG9yIGJyb3dzZXIgZ2xvYmFscwogICAgLy8gQmFzZWQgb24gaHR0cHM6Ly9naXRodWIuY29tL0ZvcmJlc0xpbmRlc2F5L3VtZC9ibG9iL21hc3Rlci90ZW1wbGF0ZS5qcwogICAgdmFyIG1haW5FeHBvcnRzID0gbmV3UmVxdWlyZShlbnRyeVtlbnRyeS5sZW5ndGggLSAxXSk7CgogICAgLy8gQ29tbW9uSlMKICAgIGlmICh0eXBlb2YgZXhwb3J0cyA9PT0gJ29iamVjdCcgJiYgdHlwZW9mIG1vZHVsZSAhPT0gJ3VuZGVmaW5lZCcpIHsKICAgICAgbW9kdWxlLmV4cG9ydHMgPSBtYWluRXhwb3J0czsKCiAgICAgIC8vIFJlcXVpcmVKUwogICAgfSBlbHNlIGlmICh0eXBlb2YgZGVmaW5lID09PSAnZnVuY3Rpb24nICYmIGRlZmluZS5hbWQpIHsKICAgICAgZGVmaW5lKGZ1bmN0aW9uKCkgewogICAgICAgIHJldHVybiBtYWluRXhwb3J0czsKICAgICAgfSk7CgogICAgICAvLyA8c2NyaXB0PgogICAgfSBlbHNlIGlmIChnbG9iYWxOYW1lKSB7CiAgICAgIHRoaXNbZ2xvYmFsTmFtZV0gPSBtYWluRXhwb3J0czsKICAgIH0KICB9CgogIC8vIE92ZXJyaWRlIHRoZSBjdXJyZW50IHJlcXVpcmUgd2l0aCB0aGlzIG5ldyBvbmUKICByZXR1cm4gbmV3UmVxdWlyZTsKfSh7IjBiN2M5OTNlY2RlODZiNDJhY2NkYjNiNzlkZTY2ZTEyIjpbZnVuY3Rpb24ocmVxdWlyZSxtb2R1bGUsZXhwb3J0cykgewpjb25zdCBsb2NhbCA9IHJlcXVpcmUoJy4vbG9jYWwnKTsKCm1vZHVsZS5leHBvcnRzID0gZnVuY3Rpb24gKCkgewogIHJldHVybiBsb2NhbC5hICsgbG9jYWwuYjsKfTsKfSx7Ii4vbG9jYWwiOiI4YzNiOWQ3ZTI3YTM2MzU5OGJiMDNmZTQ2NWNhZmQ5NCJ9XSwiOGMzYjlkN2UyN2EzNjM1OThiYjAzZmU0NjVjYWZkOTQiOltmdW5jdGlvbihyZXF1aXJlLG1vZHVsZSxleHBvcnRzKSB7CmNvbnN0IHV0aWwgPSByZXF1aXJlKCcuL3V0aWxzL3V0aWwnKTsKCmV4cG9ydHMuYSA9IDU7CmV4cG9ydHMuYiA9IHV0aWwuY291bnQoNCwgNSk7Cn0seyIuL3V0aWxzL3V0aWwiOiIyNDc1YzA2ODg0MmZkZmY4NWFjODUxYWU1Mjg2Yzc5ZiJ9XSwiMjQ3NWMwNjg4NDJmZGZmODVhYzg1MWFlNTI4NmM3OWYiOltmdW5jdGlvbihyZXF1aXJlLG1vZHVsZSxleHBvcnRzKSB7CmV4cG9ydHMuY291bnQgPSBmdW5jdGlvbiAoYSwgYikgewogIHJldHVybiBhICsgYjsKfTsKfSx7fV19LHt9LFsiMGI3Yzk5M2VjZGU4NmI0MmFjY2RiM2I3OWRlNjZlMTIiXSwgbnVsbCkKCi8vIyBzb3VyY2VNYXBwaW5nVVJMPWluZGV4LmpzLm1hcAo=,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInBhY2thZ2VzL2NvcmUvaW50ZWdyYXRpb24tdGVzdHMvdGVzdC9pbnRlZ3JhdGlvbi9zb3VyY2VtYXAtbmVzdGVkL2luZGV4LmpzIiwicGFja2FnZXMvY29yZS9pbnRlZ3JhdGlvbi10ZXN0cy90ZXN0L2ludGVncmF0aW9uL3NvdXJjZW1hcC1uZXN0ZWQvbG9jYWwuanMiLCJwYWNrYWdlcy9jb3JlL2ludGVncmF0aW9uLXRlc3RzL3Rlc3QvaW50ZWdyYXRpb24vc291cmNlbWFwLW5lc3RlZC91dGlscy91dGlsLmpzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiI7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7Ozs7OztBQUFBO0FBQ0E7QUFDQTtBQUNBO0FBQ0E7O0FDSkE7QUFDQTtBQUNBO0FBQ0E7O0FDSEE7QUFDQTtBQUNBIiwiZmlsZSI6ImluZGV4LmpzLm1hcCIsInNvdXJjZVJvb3QiOiIuLi8uLi8uLi8uLi8uLi8uLi8uLi8ifQ==,Y29uc3QgbG9jYWwgPSByZXF1aXJlKCcuL2xvY2FsJyk7Cgptb2R1bGUuZXhwb3J0cyA9IGZ1bmN0aW9uKCkgewogIHJldHVybiBsb2NhbC5hICsgbG9jYWwuYjsKfQ==,Y29uc3QgdXRpbCA9IHJlcXVpcmUoJy4vdXRpbHMvdXRpbCcpOwoKZXhwb3J0cy5hID0gNTsKZXhwb3J0cy5iID0gdXRpbC5jb3VudCg0LCA1KTs=,ZXhwb3J0cy5jb3VudCA9IGZ1bmN0aW9uKGEsIGIpIHsKICByZXR1cm4gYSArIGI7Cn0=)

**Pull request merged**